### PR TITLE
ci: Add Workflow for ChromaticUI Visual Regression Testing

### PR DESCRIPTION
**Pull request by TylerAPfledderer**
## Description

This PR adds the Github action to automate Chromatic UI's visual regression testing in conjunction with StorybookJS. 

The workflow is customized for the specific usage of this project with the following expectations:
- Action to be triggered only if there are file changes within the `components`, `pages`, `pages-conditional`, and `templates` directories (subject to be altered later as needed)
- Testing only fails if errors exist in story files.
- Reliance on ensuring that the steps generated by this action in the PR checks are to be reviewed and accepted before merge, and that if any change is denied within the Chromatic dashboard by a reviewer, that it should fail the check and prevent the PR from being merged.

Along with comments given by me to note anything specific in the file or questions, please provide further feedback on any additions or alterations.

## Related issue

#8490

### CODE REVIEW

File: .github/workflows/chromatic.yml [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    types:
+      - opened
+      - synchronize
+      - ready_for_review
+    paths:
+      # Only run on file changes in any of these paths
+      - "src/components/**/*"
+      - "src/pages/**/*"
+      - "src/pages-conditional/**/*"
+      - "src/templates/**/*"
```
Comment (TylerAPfledderer) [inline .github/workflows/chromatic.yml:18]: Only trigger in a PR that is opened, has new commits or related changes, or is converted from draft to "Ready for Review".

Also only runs if any files in the given paths have changed.

File: .github/workflows/chromatic.yml [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  chromatic-deployment:
+    # Operating System
+    runs-on: ubuntu-latest
+    # Job steps
+    steps:
+      # 👇 Version 2 of the action
+      - name: Checkout repo
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: 0 # 👈 Required to retrieve git history
```
Comment (TylerAPfledderer) [inline .github/workflows/chromatic.yml:31]: See Chromatic docs on [Support for actions/checkout@v2 and above](https://www.chromatic.com/docs/github-actions#support-for-codeactionscheckoutv2code-and-above)

File: .github/workflows/chromatic.yml [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        # 👇 Install dependencies with the same package manager used in the project (replace it as needed), e.g. yarn, npm, pnpm
+        run: yarn
+      - name: Publish to Chromatic
+        # 👇 Adds Chromatic as a step in the workflow
+        uses: chromaui/action@v1
+        # Options required for Chromatic's GitHub Action
+        with:
+          projectToken: ${{ secrets.CHROMATIC_PROJECT_TOKEN }}
+          # 👇 Only fail if Storybook contains stories that error
+          exitZeroOnChanges: true
```
Comment (TylerAPfledderer) [inline .github/workflows/chromatic.yml:42]: Along with these options, `onlyChanged` can be considered in the future. This enables Chromatic's TurboSnap to speed up the testing process by only updating snapshots of stories related to the files that have changed.

Can be an alternative or an addition to [lines 13 to 18](https://github.com/ethereum/ethereum-org-website/pull/9234/files#r1080664656)

See [TurboSnap](https://www.chromatic.com/docs/turbosnap)

**Comment by TylerAPfledderer** [comment]
@pettinarip keeping in draft for now, but pinging you for visual if you would like to look it over! 

Unless you would like to go ahead and mark it "Ready for Review" 😸

File: .github/workflows/chromatic.yml (hunk 3 above)
Comment (pettinarip) [reply .github/workflows/chromatic.yml:42]: Cool. Yeah, it seems they don't let you enable it immediately when you start a project anyway.

Agree with you. To start with, I think we're good with the current setup.

**Review by pettinarip — COMMENTED** [review]
@TylerAPfledderer nice! this looks good for me. I don't see that we need to add any other change to this setup.

Once we merge this, what do you see as next step? 

Should I manually push the `dev` branch with `npx chromatic --project-token={...}` to complete the setup on Chromatic side?

**Comment by TylerAPfledderer** [comment]
> Should I manually push the dev branch with npx chromatic --project-token={...} to complete the setup on Chromatic side?

That would be the next step to establish the connection.

Once the script is run, you should be able to see those "test" stories put into the dashboard library.

**Comment by TylerAPfledderer** [comment]
Based on a test I did for a personal project, scripting for Chromatic and/or adding it as a devDep do not look to be needed. Three things should occur in a PR:
1. The run of the Action
2. Checks appear for UI Test, UI Review, and a published Storybook instance
    - Going into the "details" of the UI Tests goes to the Chromatic dashboard to review all changed component and approve/deny the visual changes. If any one of the changes are denied, the UI Test check fails. See [UI Test Docs](https://www.chromatic.com/docs/test)
    - Going into the "details" of the UI Review is to approve the entire PR of the visual changes. It gives you the ability to invite other reviewers for this approval if necessary before merging. See [UI Review Docs](https://www.chromatic.com/docs/review)

**Comment by pettinarip** [comment]
Got it. Thanks.

Yea, seems that the `chromatic` dep is only useful to do the manual or local push but since we are not going to do that, its not necessary (you could also run it with `npx chromatic` either way).

I'll do the initial push and see how it goes 🤞🏼 will keep you posted.

**Comment by pettinarip** [comment]
@TylerAPfledderer I think we are ready for this. I've pushed `dev` to the project.

One question, do you know at what step of the process we need to install the [Chromatic Github App](https://github.com/apps/chromatic-com)? I think is the one that is going to create our status checks on every PR.

Its the only missing part in my head.

**Comment by TylerAPfledderer** [comment]
The app is for enabling permissions in the "UI Review" check when inviting others to review after approving all the changes under the "UI Tests" check.

[UI Review docs - Enable](https://www.chromatic.com/docs/review#enable)

So it looks like this app can be added at any time before or during the next PR with Storybook 😄

**Comment by TylerAPfledderer** [comment]
@pettinarip circling back to this to see if this good to bring online! 😄 Would be nice to test out if you are good with adding stories to the existing components (minus those to be consolidated or removed) for the visual testing.

**Comment by pettinarip** [comment]
> @pettinarip circling back to this to see if this good to bring online! smile Would be nice to test out if you are good with adding stories to the existing components (minus those to be consolidated or removed) for the visual testing.

I wanted to merge this along with the SB v7 upgrade. In order to start creating stories with the latest format. Not a big deal though but I wanted to prevent someone from creating the old ones

**Comment by TylerAPfledderer** [comment]
@pettinarip that's fine then! No problem waiting here. 😁

File: .github/workflows/chromatic.yml (hunk 1 above)
Comment (TylerAPfledderer) [reply .github/workflows/chromatic.yml:18]: @pettinarip I should probably add the path to the theme config to allow for review of visual changes if any values are altered. What do you think?

Comment (pettinarip) [reply .github/workflows/chromatic.yml:18]: Right! just seeing this now, sorry. I agree with that. Will add it.

**Review by pettinarip — APPROVED** [review]
🚀

**Pull request merged**